### ISSUE TITLE
[MSHARED-787] Add optional buildEnvironment information to the manifest

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-archiver#217]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** opened **[MSHARED-787](https://issues.apache.org/jira/browse/MSHARED-787?redirect=false)** and commented

People are demanding to have more control about the build information (MSHARED-362). We should add the following to control this: `<addBuildEnvironmentEntries` /> which is by default disabled. It will contain the folling properties (akin to `mvn- v`):

```
Build-Tool: ${maven.build.version:-Apache Maven}
Build-Jdk: ${java.version} (${java.vendor})
Build-Os:  ${os.name} (${os.version}; (${os.arch})
```

---

**Affects:** maven-archiver-3.3.0

**Issue Links:**
- [MSHARED-796](https://issues.apache.org/jira/browse/MSHARED-796) use java.specification.version instead of java.version in Build-Jdk manifest entry

- [MNG-6276](https://issues.apache.org/jira/browse/MNG-6276) Support Reproducible Builds (https://reproducible-builds.org)

- [MSHARED-799](https://issues.apache.org/jira/browse/MSHARED-799) Change "Created-By" manifest entry value to be reproducible

- [MSHARED-781](https://issues.apache.org/jira/browse/MSHARED-781) Upgrade to Plexus Archiver 4.1.0
 (_**"requires"**_)
- [MSHARED-362](https://issues.apache.org/jira/browse/MSHARED-362) Support removing default manifest entries with Maven Archiver
 (_**"is required by"**_)

**Remote Links:**
- [GitHub Pull Request #3
](https://github.com/apache/maven-archiver/pull/3)
- [Example Config
](http://andresalmiray.com/customize-jar-manifest-entries-with-maven-gradle/)
- [Upstream Issue Plexus Archiver
](https://github.com/codehaus-plexus/plexus-archiver/issues/110)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I have now pushed a branch which contains all necessary changes, but the documentation. What needs to be changed in Plexus Archiver not to add `Created-By: Plexus Archiver ${version}`.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I have pushed a new Plexus Archiver branch solving this issue and updated this branch accordingly. All should be fine now. If all participants agree, I'll fix Plexus Archiver first and then I can resolve this issue.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

If I don't hear any objections by the end of next week, I'd go on and merge this.

**Comment by jira-importer** [comment]
**[Eduardo Pérez Ureta](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=edpeur)** commented

Is there any reason to keep Created-By: Plexus Archiver ${version} ?
Should it be removed altogether?

What is the reason to keep:
Build-Os:  ${os.name} (${os.version}; (${os.arch})
?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Eduardo Pérez Ureta](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=edpeur), you obviously did not check what I have done.

* I have removed that entry from Plexus Archiver: https://github.com/codehaus-plexus/plexus-archiver/commit/b4674dceb0cc737b3623619aed182fa39c697cb4
* The build information is completely optional, and not enabled by default.

**Comment by jira-importer** [comment]
**[Eduardo Pérez Ureta](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=edpeur)** commented

You seem to have made it optional, instead of removing it.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I don't understand your problem. If you don't want it, just don't use it.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

the idea currently is to produce a separate buildinfo file instead of putting everything in the jar: https://reproducible-builds.org/docs/jvm/

with the buildinfo in place for people who want reproducible builds, I'm not convinced the feature in the manifest is really useful

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I think that many people won't care for the additional file. They be fine with the info in the manifest. If they don' need it at all, they won't turn it on. I believe that `java.version` is a better value than `java.specification.version` because the source is tested also. It can happen that code fals on 1.8.120, but work on 1.8.202 because of a fix.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

what is important to me is that by default we have only java.specification.version
but if people configure it, they can get very detailed info like java.version, os details and so on in Manifest if they want: in Manifest, this will be "the non-reproducible way", while with buildinfo, it will be "the reproducible way": it's good to have both options available and people choose their preferred option

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

By

> while with buildinfo, it will be "the reproducible way"

you mean that this information will be held out of band (outside of the JAR)?

My problem isn't having `java.specification.version` by default, but what benefit gives it to the user when you don't have the `Build-Maven-Version` also. MSHARED-362 would still remain open :-(

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy), it is worth (in terms of consistency) to rename `Created-By` to `Build-Maven`: `${maven.version&#x7d;` (`${commit-id&#x7d;`)?

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

"Created-By" is defined by Plexus Archiver: it's only overridden in maven-archiver to provide a better value than the "Plexus Archiver" value defined by Plexus Archiver

On the proposed value, while at it, consistency with "mvn -v" output can be interesting: sometimes, the build is done with a SNAPSHOT...

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

to me, "Created By: Plexus Archiver \<version of Plexus Archiver>" is reproducible
then for MSHARED-362 either we keep this default value, either we remove this entry in maven-archiver by default

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy), the `Created-By` can be disabled in Plexus Archiver now, in fact. I have disabled it in 4.0.1 (not by default) along with this branch. Since Plexus Archiver seems to in interesting. Why not say:

```
Build-Tool: Apache Maven ${maven.version} (${commit-id})
Build-Archiver: Plexus Archiver ${pa.version}
Build-Jdk: ${java.version} (${java.vm.vendor})
Build-Os:  ${os.name} (${os.version}; (${os.arch})
```

The prefix is completely consistent. Note: `Build-Jdk` will overwrite the value you have changed recently.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

What I was also considering is to have a `maven.vendor` property, because a lot of Linux distro modify Maven which is not the canonical anymore. Our value would be `ASF`, those who did not modify our binary releases can retain it, other can say `Debian`, `Fedora`, etc.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

I like the whole approach

On `Build-Archiver`, we can easily put `Maven Archiver ${ma.version`}, and ideally, this would be overridden by the plugin which did the archive: and this could even be kept as default manifest entry, since reproducible!!!

on `maven.vendor` property, this could be a good idea, but it is something that goes beyond this Jira issue: it's something for Maven core and build instructions...

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

So, your proposal is:

```
Created-By: Maven Archiver ${ma.version}
Build-Tool: Apache Maven ${maven.version} (${commit-id})
Build-Jdk: ${java.version} (${java.vm.vendor})
Build-Os:  ${os.name} (${os.version}; (${os.arch})
```

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

almost:
1. as optional build environment information (= the current MSHARED-787 issue), put:

```
Build-Tool: Apache Maven ${maven.version} (${commit-id})
Build-Jdk: ${java.version} (${java.vm.vendor})
Build-Os:  ${os.name} (${os.version}; (${os.arch})
```

2. keep `Created-By` in the default manifest but change the value to

```
Created-By: Maven Archiver ${ma.version}
```

as a separate issue (since it would be there by default, then not related to optional build environment information)

3. improve the API for calling plugins to be able to easily replace the default `Maven Archiver ${ma.version`} value with `Maven xxx Plugin ${pluginversion`}

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Agreed.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy), I have now pushed an updated version of the branch. Please have a look.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

MSHARED-799 issue created for the `Created-By` entry value change

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [93f5ba12836b7b5bb0f9de27ede8bcbd6e46551a](https://gitbox.apache.org/repos/asf?p=maven-archiver.git;a=commit;h=93f5ba12836b7b5bb0f9de27ede8bcbd6e46551a).
